_posts/pl/2021-04-26-zarzadzanie-dostepem-przy-uzyciu-access-control-list.md

```diff
@@ -0,0 +1,324 @@
+
+Kilka lat temu, kiedy zaczynałem przygodę z programowaniem, moim pierwszym komercyjnym projektem do wykonania, było zaimplementowanie aplikacji internetowej, która miała służyć do zarządzania zadaniami. Zadania te były przypisane do konkretnych użytkowników, którzy nie mogli sobie nawzajem przeglądać zadań. Problem wydawał się być powszechny. Po krótkim poszukiwaniu, natknąłem się na koncepcję znaną jako **Access Control List** oraz jego realizacją w **Spring Security ACL**, która jest rozszerzeniem Spring Security. To było to, czego szukałem!
+
+W tym artykule wyjaśnię, dlaczego Spring Security jest niewystarczający do zrealizowania wspomnianego wymagania, i dlaczego potrzebujemy rozszerzenia Spring Security ACL. Dodatkowo przedstawię fragmenty kodu, które są istotne w naszym projekcie.
+
+Pełny kod jest dostępny pod adresem [**https://github.com/pawelwalaszek/spring-security-acl**](https://github.com/pawelwalaszek/spring-security-acl).
+
+## Czym właściwie jest Access Control List?
+
+Krótka definicja brzmi:
+
+```
+Access Control List (ACL) jest listą uprawnień skojarzonych z obiektem.
+```
+
+W naszym przypadku obiektem jest zadanie. Natomiast lista uprawnień jest przechowywana w bazie danych, w specjalnych strukturach tabelarycznych, w której są zdefiniowane relacje między obiektem, a użytkownikiem.
+
+## Dlaczego Spring Security jest niewystarczający?
+
+**Spring Security** pozwala określić *dostęp na poziomie żądania HTTP lub wywołania metody*.
+
+Przykład:
+
+```java
+@PreAuthorize("hasRole('TASK')")
+public List<Task> getTasksWithoutAcl() {
+    return taskRepository.findAll();
+}
+```
+
+W powyższym przykładzie użytkownik z rolą *TASK* otrzyma pełną listę obiektów *Task*. Nie jest to zgodne z naszym wymaganiem, ponieważ chcemy, aby użytkownik otrzymał wyselekcjonwaną listę obiektów *Task*, dokładniej, listę obiektów do których został przypisany.
+
+**Spring Security ACL** pozwala określić *dostęp na poziomie obiektów*.
+
+Przykład:
+
+```java
+@PreAuthorize("hasRole('TASK')")
+@PostFilter("hasPermission(filterObject, 'READ')")
+public List<Task> getTasksWithAcl() {
+    return taskRepository.findAll();
+}
+```
+
+W powyższym przykładzie użytkownik z rolą *TASK* otrzyma listę obiektów *Task*, ale tylko do tych, do których otrzymał uprawnienie odczytu.
+
+**@PreAuthorize** – sprawdza, czy użytkownik posiada rolę *TASK*. Jeśli nie posiada, to generuje wyjątek, który tworzy odpowiedź HTTP ze statusem 403.
+
+**@PostFilter** – usuwa obiekty z kolekcji, do których użytkownik nie ma uprawnień.
+
+Kombinacja adnotacji **@PreAuthorize** i **@PostFilter** jest bardzo wygodna, ale żeby taka była, należy odpowiednio przygotować konfigurację w naszej aplikacji zarówno dla Spring Security jak i dla Spring Security ACL.
+
+## Konfiguracja
+
+a) Konfiguracja dla Spring Security:
```

Comment:
zawsze mam problem z pozostawianiem takiej ściany kody bez komentarza. W szczególności część tego kodu nie jest istotna dla opisywanej sytuacji, jak na przykład cachowanie. Bardziej do mnie przemawia takie przeplatanie kodu konfiguracji springowej z krótkimi komentarzami. Jak na przykład tutaj: https://www.baeldung.com/spring-security-acl. 
Nie trzeba  tutaj we wpisie umieszczać całych klas, tymbardziej że dajesz odnośnik do githuba.

Reply:
Poprawione